credentials/tls_ext_test.go

```diff
@@ -236,3 +238,72 @@ func (s) TestTLS_CipherSuitesOverridable(t *testing.T) {
 		t.Fatalf("EmptyCall err = %v; want <nil>", err)
 	}
 }
+
+// TestTLS_DisabledALPN tests the behaviour of a gRPC client when connecting to
+// a server that doesn't support ALPN.
+func (s) TestTLS_DisabledALPN(t *testing.T) {
+	initialVal := envconfig.EnforceALPNEnabled
+	defer func() {
+		envconfig.EnforceALPNEnabled = initialVal
+	}()
+
+	serverCfg := &tls.Config{
+		Certificates: []tls.Certificate{serverCert},
+		NextProtos:   []string{}, // Empty list indicates ALPN is disabled.
+	}
+
+	tests := []struct {
+		name         string
+		alpnEnforced bool
+		wantErr      bool
+	}{
+		{
+			name:         "enforced",
+			alpnEnforced: true,
+			wantErr:      true,
+		},
+		{
+			name: "not_enforced",
+		},
+	}
+
+	for _, tc := range tests {
+		t.Run(tc.name, func(t *testing.T) {
+			envconfig.EnforceALPNEnabled = tc.alpnEnforced
+
+			listner, err := tls.Listen("tcp", "localhost:0", serverCfg)
+			if err != nil {
+				t.Fatalf("Error starting TLS server: %v", err)
+			}
+
+			go func() {
+				conn, err := listner.Accept()
+				if err != nil {
+					t.Errorf("tls.Accept failed err = %v", err)
+				} else {
+					conn.Write([]byte("Hello, World!"))
```

Comment:
What purpose does this call to `Write` serve? Can we get rid of it?

Reply:
Since this a tls listner, it automatically does a TLS handshake when writing for the first time. An alternate way is to cast the net.Listener into a tls.Listener and call it's handshake method.

Reply:
Just a one liner comment to say that this call to `Write` is to ensure that the underlying handshake happens would be helpful for a future reader of this code. Thanks.

Reply:
Added a comment.